src/Silk/Services/Guild/AutoMod/RaidDetectionService.cs

```diff
@@ -0,0 +1,223 @@
+            return Result.FromSuccess();
+        
+        TrimMessageBuckets(); // Tidy up our buckets before we check if the server even has raid detection enabled.
+
+        var config = await _config.GetConfigAsync(guildID);
+        
+        if (!config.EnableRaidDetection)
+            return Result.FromSuccess();
+        
+        var bucket = _messageBuckets.GetOrAdd(channelID, _ => new());
+        
+        bucket.Add(new MessageDTO(guildID, channelID, messageID, AuthorID, content.GetHashCode()));
+        
+        var groups = bucket.GroupBy(x => x.HashCode);
+
+        foreach (var group in groups)
+        {
+            if (group.Count() < 3)
+                continue;
+
+            if (DateTimeOffset.UtcNow - group.Last().MessageID.Timestamp > TimeSpan.FromSeconds(15))
+                continue; // Given up? Or just not a raid. This is an abitrary decision.
+            
+            foreach (var raider in group)
+                await _raiders.Writer.WriteAsync(new Raider(raider.GuildID, raider.AuthorID, "Suspected raid: Coordinated message raid detected."));
+            
+            bucket.RemoveAll(r => group.Contains(r));
+        }
+        
+        return Result.FromSuccess();
+    }
+
+    public async Task<Result> HandleJoinAsync(Snowflake GuildID, Snowflake UserID, DateTimeOffset joined, bool hasAvatar)
+    {
+        TrimBuckets(GuildID);
+
+        var config = await _config.GetConfigAsync(GuildID);
+        
+        if (!config.EnableRaidDetection)
+            return Result.FromSuccess();
+
+        var joinCounter = _joins.GetOrAdd(GuildID, _ => new());
+
+        if (DateTimeOffset.UtcNow - joinCounter.LastJoin > TimeSpan.FromSeconds(config.RaidCooldownSeconds))
+        {
+            joinCounter.Count = 0;
+        }
+        
+        if (joinCounter.Count++ >= config.RaidDetectionThreshold)
+        {
+            await _raiders.Writer.WriteAsync(new Raider(GuildID, UserID, "Raid Detection: Join velocity check exceeds threshold."));
+
+            return Result.FromSuccess();
+        }
+        
+        joinCounter.LastJoin = joined;
+        _joins[GuildID] = joinCounter;
+        
+        // Join velocity check failed. Check for chunked raid activity.
+        var joinRanges = _joinRanges.GetOrAdd(GuildID, new List<JoinRange>());
+
+        var created = UserID.Timestamp;
+        
+        var range = joinRanges.FirstOrDefault(r => r.InRange(created));
+
+        if (range is null)
+        {
+            range = new(created.AddHours(-2), created.AddHours(.5), config.RaidDetectionThreshold);
+            joinRanges.Add(range);
+        }
+
+        if (range.IsSuspect(created))
+        {
+            await _raiders.Writer.WriteAsync(new Raider(GuildID, UserID, "Raid Detection: Join cluster check failed."));
+            return Result.FromSuccess();
+        }
+        
+        //TODO: Check for avatar-less users & calculate risk factor based on account date
```

Comment:
Don't forget about the TODO

Reply:
This is more of a "Future me TODO"; it's something that'll be accounted for in V2.